Bug 1829445: Fixes: Container image does not validate dynamically

sahil143 opened the pull request:
**Fixes**: https://issues.redhat.com/browse/ODC-3178
<!-- For e.g Fixes: https://issues.redhat.com/browse/ODC-XXX -->

**Analysis / Root cause**: 
<!-- Briefly describe analysis of US/Task or root cause of Defect -->
container image validation only happens on blur, which caused some issue

**Solution Description**: 
<!-- Describe your code changes in detail and explain the solution -->
enable image validation while user types

**Screen shots / Gifs for design review**: 
<!-- If change affects UI in any way, tag @openshift/team-devconsole-ux and add screenshots/gifs  -->
![container-image](https://user-images.githubusercontent.com/9278015/80608073-e9580580-8a53-11ea-9e65-af1ba3c96d32.gif)


**Unit test coverage report**: 
<!-- Attach test coverage report -->

**Test setup:**
<!-- If any setup required to test this PR, mention the details -->

**Browser conformance**: 
<!-- To mark tested browsers, use [x] -->
- [ ] Chrome
- [ ] Firefox
- [ ] Safari
- [ ] Edge

invincibleJai commented on frontend/packages/dev-console/src/components/import/image-search/ImageSearch.tsx:
```diff
@@ -30,85 +30,79 @@ const ImageSearch: React.FC = () => {
-              kind: 'DockerImage',
-              name: _.trim(searchTermImage),
+  const handleSearch = React.useCallback(
+    _.debounce((newValues: any) => {
```
nit: type here

rohitkrai03 commented on frontend/packages/dev-console/src/components/import/image-search/ImageSearch.tsx:
```diff
@@ -151,8 +136,8 @@ const ImageSearch: React.FC = () => {
         validated={validated}
-        onBlur={() => {
-          handleSearch();
+        onKeyUp={() => {
```
Why not use `onChange` here?

sahil143 replied:
fixed

rohitkrai03 commented on frontend/packages/dev-console/src/components/import/image-search/ImageSearch.tsx:
```diff
@@ -125,10 +120,6 @@ const ImageSearch: React.FC = () => {
     <span>{values.searchTerm === '' ? 'Required' : values.isi.status?.message}</span>
   );
 
-  React.useEffect(() => {
```
This would break edit application scenario. We need this logic here.

rohitkrai03 commented on frontend/packages/dev-console/src/components/import/image-search/ImageSearch.tsx:
```diff
@@ -151,8 +142,8 @@ const ImageSearch: React.FC = () => {
-          handleSearch();
+        onChange={(e: KeyboardEvent) => {
+          handleSearch((e.target as HTMLInputElement).value);
           setFieldTouched('searchTerm', true);
```
Pretty sure you don't need to do that now since default `onBlur` on the field will handle this.

rohitkrai03 commented on frontend/packages/dev-console/src/components/import/image-search/ImageSearch.tsx:
```diff
@@ -17,98 +17,93 @@ import { secretModalLauncher } from '../CreateSecretModal';
 
   const handleSave = (name: string) => {
     setNewImageSecret(name);
-    values.searchTerm && handleSearch();
```
I think this one is being used to trigger the validation again when a pull secret is added. So we need this one as well.

rohitkrai03 commented:
/assign

christianvogt commented:
The gif looks odd to me.
Firstly you are typing fairly quick and yet it still started validating.

Secondly as soon as the user inputs a value which no longer matches that which is currently validating or that which has already errored I think that the helper text should be cleared immediately.

Please involve UX for approval.

sahil143 commented:
/cc @serenamarie125 @beaumorley @openshift/team-devconsole-ux

sahil143 commented:
updated Gif
![validate-2](https://user-images.githubusercontent.com/9278015/81165281-5ec15a00-8faf-11ea-83dc-292b9f2a22d9.gif)

christianvogt commented:
/lgtm cancel

I can submit the form after immediately changing the image name to something that is invalid. The form submits with the previous valid value.

The field state should be immediately invalidated on change of the image name value.

sahil143 commented:
> /lgtm cancel
> 
> I can submit the form after immediately changing the image name to something that is invalid. The form submits with the previous valid value.
> 
> The field state should be immediately invalidated on change of the image name value.

Fixed. @christianvogt Please review again

christianvogt commented:
@sahil143 while the create button is properly disabled, the helper text remains while typing. This helper text should also be cleared.
![image](https://user-images.githubusercontent.com/14068621/81318587-7f93b900-905c-11ea-9e54-13a45567612f.png)

sahil143 commented:
> @sahil143 while the create button is properly disabled, the helper text remains while typing. This helper text should also be cleared.
> ![image](https://user-images.githubusercontent.com/14068621/81318587-7f93b900-905c-11ea-9e54-13a45567612f.png)

fixed!

christianvogt commented on frontend/packages/dev-console/src/components/import/image-search/ImageSearch.tsx:
```diff
@@ -151,9 +151,10 @@ const ImageSearch: React.FC = () => {
-          handleSearch();
-          setFieldTouched('searchTerm', true);
+        onChange={(e: KeyboardEvent) => {
+          setFieldValue('disableSubmit', true);
```
I really don't think this field is necessary. We were able to have a disabled submit button without it. You should be resetting the values previously set by the error or success conditions. See what is done in the `catch` of the search.

sahil143 replied:
disabled submit button is based on only two props i.e. `dirty` and `errors`. Not sure how we can disable the button based on these and in catch block we are only setting the image stream data to empty values and error.

christianvogt replied:
The submit button disabled state is controlled by the validation of the form data. But deleting the required image values the form will be in an invalid state, thereby disabling the submit button.

By not having a valid image, the submit button will be disabled. So setting the values back to empty (as they were on initializing the form), we get the desired outcome.

karthikjeeyar replied:
We can remove the `disableSubmit` prop from everywhere and  here we can set  `setFieldValue('isi', {});` so the form will be disabled if there is no valid image in the isi

sahil143 replied:
fixed!

sahil143 commented:
>I also noticed that after entering a valid image, then changing to a new valid image, the old app name and name values remain which no longer coincide with the new image. They should be modified again if the user never touched those fields. I'm ok if we tackle this in a new issue though

fixed! added a method to reset the fields based on form type

abhi-kn commented:
/bugzilla refresh

christianvogt commented:
/lgtm
/approve
